uvicorn/protocols/http/httptools_impl.py

```diff
@@ -244,7 +256,7 @@ def on_headers_complete(self) -> None:
         self.scope["method"] = method.decode("ascii")
         if http_version != "1.1":
             self.scope["http_version"] = http_version
-        if self.parser.should_upgrade():
+        if self.parser.should_upgrade() and self._should_upgrade():
```

Comment:
Why is it an "and"? :thinking:

Reply:
The C implementation of parser.should_upgrade looks like this:
```
uint8_t llhttp_get_upgrade(llhttp_t* parser) {
  return parser->upgrade;
}
```
And the flag is set during parsing anyways, so this is faster than our iterate over headers implementation. If the parser didn't detect an upgrade request we don't need to do our checks. If the parser detected an upgrade we still need to decide if we want/can upgrade.
So the first part is functionally not needed but should perform better